feat: add comprehensive firewall custom organization policy constraints

riddopic opened the pull request:
## Summary

This PR adds a comprehensive set of firewall custom organization policy constraints to enforce security best practices for VPC firewall rules and firewall policies in Google Cloud.

### New Firewall Constraints

**VPC Firewall Rules:**
- `firewallRestrictDatabasesRule` - Restrict ingress access to database ports (MySQL, PostgreSQL, MSSQL, Oracle, MongoDB, Cassandra, Redis, Memcached, CouchDB)
- `firewallRestrictDirectoryServicesRule` - Restrict ingress access to directory service ports (LDAP, Kerberos, SMB)
- `firewallRestrictInsecureProtocolsRule` - Restrict ingress access to insecure protocols (Telnet, FTP, TFTP, SNMPv1/v2)
- `firewallRestrictMailProtocolsRule` - Restrict ingress access to mail protocols (SMTP, POP3, IMAP)
- `firewallRestrictManagementPortsRule` - Restrict ingress access to management ports (RDP, SSH, WinRM, VNC)
- `firewallRestrictNetworkServicesRule` - Restrict ingress access to network services (DNS, DHCP, NTP, SNMP)
- `firewallRestrictOpenFirewallRule` - Prevent overly permissive firewall rules (0.0.0.0/0 or ::/0)
- `firewallEnforceRuleLogging` - Require logging to be enabled on firewall rules

**Firewall Policy Rules:**
- `firewallRestrictDatabasesPolicyRule` - Restrict ingress access to database ports in firewall policies
- `firewallRestrictDirectoryServicesPolicyRule` - Restrict ingress access to directory services in firewall policies
- `firewallRestrictInsecureProtocolsPolicyRule` - Restrict insecure protocols in firewall policies
- `firewallRestrictMailProtocolsPolicyRule` - Restrict mail protocols in firewall policies
- `firewallRestrictManagementPortsPolicyRule` - Restrict management ports in firewall policies
- `firewallRestrictNetworkServicesPolicyRule` - Restrict network services in firewall policies
- `firewallRestrictOpenFirewallPolicyRule` - Prevent overly permissive firewall policy rules
- `firewallEnforcePolicyRuleLogging` - Require logging on firewall policy rules

### Other Changes

- Enhanced `firewallEnforceNamingConvention` and `firewallRequireDescription` constraints
- Added `iamDisableRedisAdminRoles` constraint to restrict Redis admin role assignments
- Added `accesscontextmanagerDisableBridgePerimeters` constraint
- Added `cloudsqlRequirePostgreSQLDatabaseAdditionalFlags` constraint
- Updated schema definitions for firewall, IAM, CloudSQL, and Access Context Manager services
- Comprehensive test coverage for all new constraints

### Test Plan

- [x] All new constraints have corresponding test cases
- [x] Test cases cover both allow and deny scenarios
- [x] VPC firewall rule tests in `tests/test_cases/firewall/vpc_rule_*.yaml`
- [x] Firewall policy tests in `tests/test_cases/firewall/policy_rule_*.yaml`

riddopic committed:
feat: add comprehensive firewall custom organization policy constraints

Add a comprehensive set of firewall custom organization policy constraints
to enforce security best practices for VPC firewall rules and firewall
policies in Google Cloud.

New VPC Firewall Rule Constraints:
- firewallRestrictDatabasesRule: Restrict ingress to database ports
- firewallRestrictDirectoryServicesRule: Restrict ingress to directory services
- firewallRestrictInsecureProtocolsRule: Restrict insecure protocols
- firewallRestrictMailProtocolsRule: Restrict ingress to mail protocols
- firewallRestrictManagementPortsRule: Restrict ingress to management ports
- firewallRestrictNetworkServicesRule: Restrict ingress to network services
- firewallRestrictOpenFirewallRule: Prevent overly permissive rules (0.0.0.0/0)
- firewallEnforceRuleLogging: Require logging on firewall rules

New Firewall Policy Rule Constraints:
- firewallRestrictDatabasesPolicyRule: Restrict ingress to database ports
- firewallRestrictDirectoryServicesPolicyRule: Restrict ingress to directory services
- firewallRestrictInsecureProtocolsPolicyRule: Restrict insecure protocols
- firewallRestrictMailProtocolsPolicyRule: Restrict mail protocols
- firewallRestrictManagementPortsPolicyRule: Restrict management ports
- firewallRestrictNetworkServicesPolicyRule: Restrict network services
- firewallRestrictOpenFirewallPolicyRule: Prevent overly permissive policy rules
- firewallEnforcePolicyRuleLogging: Require logging on policy rules

Additional Changes:
- Enhanced firewallEnforceNamingConvention and firewallRequireDescription
- Added iamDisableRedisAdminRoles constraint
- Added accesscontextmanagerDisableBridgePerimeters constraint
- Added cloudsqlRequirePostgreSQLDatabaseAdditionalFlags constraint
- Updated schemas for firewall, IAM, CloudSQL, and Access Context Manager
- Comprehensive test coverage for all new constraints
- Removed deprecated gkeAllowedInitialClusterVersions constraint

vannicktrinquier commented on tools/custom-organization-policy-library/build/custom-constraints/firewall/firewallRestrictDatabasesPolicyRule.yaml:
```diff
@@ -0,0 +1,30 @@
+#@   #! PostgreSQL: TCP/UDP 5432
+#@   #! Redis: TCP 6379
+#@   source_check = '!rule.match.srcIpRanges.all(ipRange, ipRange in ' + str(allowed_source_ranges) + ')'
+#@   tcp_ports = "port == '7000' || port == '7001' || port == '7199' || port == '8888' || port == '9042' || port == '9160' || port == '61620' || port == '61621' || port == '9200' || port == '9300' || port == '11211' || port == '11214' || port == '11215' || port == '27017' || port == '27018' || port == '27019' || port == '3306' || port == '1521' || port == '2483' || port == '2484' || port == '5432 ...
```
make a list instread like ['7000', '7001']

vannicktrinquier replied:
l4config.ports.exists(port, " + port in  mylist "))

vannicktrinquier commented on tools/custom-organization-policy-library/build/custom-constraints/firewall/firewallRestrictDatabasesRule.yaml:
```diff
@@ -0,0 +1,58 @@
+#@   #! Redis: TCP 6379
+#@   source_check = '!resource.sourceRanges.all(range, range in ' + str(allowed_source_ranges) + ')'
+#@   port_checks = [
+#@     "resource.allowed.containsFirewallPort('tcp', '7000')",
```
denied_tcp_ports = []
denied_udp_ports = []

resource.allowed.containsFirewallPort('tcp', on each of the list denied_tcp_ports)
resource.allowed.containsFirewallPort('udp', on each of the list denied_udp_ports)

riddopic committed:
refactor: improve firewall constraint port checks using list iteration

Replace chained || operators and multiple containsFirewallPort calls with
cleaner list-based 'in' operator pattern for better maintainability and
readability. This simplifies port checking logic across all firewall
constraints by using CEL's 'port in [list]' syntax instead of multiple
individual comparisons.

Changes:
- Consolidated port checks in Firewall Policy constraints to use 'port in [...]'
- Refactored VPC Firewall Rule constraints to use exists() with list checking
- Applies to database, directory services, mail protocols, insecure protocols,
  network services, management ports, and open firewall constraints

riddopic committed:
test: enable logging for firewall rule test cases

Add enable-logging: true to default_command_flags in all firewall test
cases to ensure firewall rules created during tests have logging enabled.
This applies to both VPC Firewall Rules and Firewall Policy Rules tests.

riddopic committed:
refactor: optimize firewall constraint condition logic

Improve port checking in firewall constraints by:
- Using sanitized_allowed_ranges to handle empty strings
- Simplifying layer4 config checks with direct protocol comparisons
- Reducing nesting depth in condition expressions

This refactoring maintains the same security enforcement while
improving readability and reducing CEL expression complexity.

riddopic committed:
feat: split database firewall rules into specific categories

Replace generic database firewall rules with three targeted constraints:
- firewallRestrictSQLDatabases: MySQL, PostgreSQL, SQL Server, Oracle
- firewallRestrictNoSQLDatabases: MongoDB, Cassandra, CouchDB
- firewallRestrictCacheSearchDatabases: Redis, Memcached, Elasticsearch

This provides more granular control over database access policies,
allowing organizations to enforce different source range restrictions
for different database types based on their security requirements.

Changes:
- Add schema definitions for three new constraint types
- Create six new constraint files (VPC and Policy rules for each type)
- Remove deprecated generic database constraint files
- Update values.yaml with configuration for new constraints

riddopic committed:
test: add test cases for split database firewall constraints

Replace generic database test cases with specific tests for:
- SQL databases (MySQL, PostgreSQL, SQL Server, Oracle)
- NoSQL databases (MongoDB, Cassandra, CouchDB)
- Cache/Search databases (Redis, Memcached, Elasticsearch)

Each category now has dedicated test cases covering both VPC firewall
rules and firewall policy rules, ensuring proper validation of the
port-specific restrictions for each database type.

riddopic committed:
sync
